File: Scripts/GravityLift.cs
using UnityEngine;
using System.Collections;
using System.Text;

public class GravityLift : MonoBehaviour {
	public float strength = 12f;
	public float offStrengthFactor = 0.3f;
	private Rigidbody otherRbody;
	private float modulatedStrengthY;
	public float distancePaddingToTopPoint = 0.32f; // Add half the player's capsule height(1f) to the top extent of the box collider
	public bool active = true;
	public Vector3 topPoint;
	public float initialBurstFinished;
	private BoxCollider boxcol;
	private static StringBuilder s1 = new StringBuilder();

	void Awake() {
		boxcol = GetComponent<BoxCollider>();
		if (boxcol == null) return;
		topPoint = (Vector3){0f,boxcol.bounds.max.y,0f);
	}

	void OnTriggerExit(Collider other) {
		if (other.gameObject.GetComponent<PlayerMovement>() != null) {
			Eng_Global->instances[PLAYER1].gravliftState = false;
		}
	}

	void OnForce(Collider other, bool initial) {
		if (other.gameObject.layer == 12) { // Player
			if (other.gameObject.GetComponent<PlayerMovement>() != null) {
				Eng_Global->instances[PLAYER1].gravliftState = true;
			}
		}

		float topY = Eng_Global->instances[i].position.y + (boxcol.size.y/2f);
		float dist = topY - other.gameObject.Eng_Global->instances[i].position.y + 0.48f;
		float velY = otherRbody.velocity.y;
		if (otherRbody.velocity.y < 0f) velY = 0f; // Saturate at bottom end.

		if (dist < distancePaddingToTopPoint) {
			Vector3 force = (Vector3){0f,9.81f - velY,0f);
			otherRbody.AddForce(force,ForceMode.Acceleration);
		} else {
			if (otherRbody.velocity.y < (strength * otherRbody.mass)) {
				float yForce = ((strength * otherRbody.mass)
								- otherRbody.velocity.y);

				if (initial
					|| initialBurstFinished > Eng_Global->pauseRelativeTime) {

					yForce *= 2f;
				}

				otherRbody.AddForce((Vector3){0f,yForce,0f));
			}
		}
	}

	void OffForce(Collider other, bool initial) {
		// Apply weak force for inactive state - applies some force for gentle
		// descent, never really off completely.
		if (other.gameObject.GetComponent<PlayerMovement>() != null) {
			Eng_Global->instances[PLAYER1].gravliftState = true;
		}

		if (otherRbody.velocity.y < offStrengthFactor) {
			float yForce = ((offStrengthFactor)-otherRbody.velocity.y);
			if (initial
				|| initialBurstFinished > Eng_Global->pauseRelativeTime) {

				yForce *= 2f;
			}

			otherRbody.AddForce((Vector3){0f,yForce,0f));
		}
	}

	void OnTriggerEnter(Collider other) {
		otherRbody = other.gameObject.GetComponent<Rigidbody>();
		if (otherRbody == null) return; // Not a physical object.

		initialBurstFinished = Eng_Global->pauseRelativeTime + 1.0f;
		if (active) OnForce(other,true);
		else OffForce(other,true);
	}

	void OnTriggerStay(Collider other) {
		otherRbody = other.gameObject.GetComponent<Rigidbody>();
		if (otherRbody == null) return; // Not a physical object.

		if (active) OnForce(other,false);
		else OffForce(other,false);
	}

	public void Toggle() {
		active = !active;
	}
}

File: Scripts/gravitylift.c
#include "mod.h"

void GravityLiftInitAfterLoad(uint16_t self) {
    Entity* e = &Eng_Global->instances[self];
    if (e->strength <= 0.0f) e->strength = 12.0f;
    if (e->offStrengthFactor <= 0.0f) e->offStrengthFactor = 0.3f;
    if (e->distancePaddingToTopPoint <= 0.0f) e->distancePaddingToTopPoint = 0.32f;
    e->topPoint = (Vector3){ 0.0f, e->position.y + (e->colliderSize.y * 0.5f), 0.0f };
}

void GravityLiftOnTriggerExit(uint16_t self, uint16_t other) {
    (void)self;
    if (other == PLAYER1) flag_set(&Eng_Global->instances[PLAYER1].entflags,ENTFLAG_GRAV_LIFT_STATE,false);
}

void GravityLiftOnForce(uint16_t self, uint16_t other, bool initial) {
    Entity* e = &Eng_Global->instances[self];
    Entity* o = &Eng_Global->instances[other];
    if (other == PLAYER1) flag_set(&Eng_Global->instances[PLAYER1].entflags,ENTFLAG_GRAV_LIFT_STATE,true);
    float topY = e->position.y + (e->colliderSize.y * 0.5f);
    float dist = topY - o->position.y + 0.48f;
    float velY = o->velocity.y < 0.0f ? 0.0f : o->velocity.y;
    if (dist < e->distancePaddingToTopPoint) AddForce(other,(Vector3){0.0f,9.81f - velY,0.0f},false); // TODO accel-vs-force parity
    else if (o->velocity.y < (e->strength * o->mass)) {
        float yForce = (e->strength * o->mass) - o->velocity.y;
        if (initial || e->initialBurstFinished > Eng_Global->pauseRelativeTime) yForce *= 2.0f;
        AddForce(other,(Vector3){0.0f,yForce,0.0f},false);
    }
}

void GravityLiftOffForce(uint16_t self, uint16_t other, bool initial) {
    Entity* e = &Eng_Global->instances[self];
    Entity* o = &Eng_Global->instances[other];
    if (other == PLAYER1) flag_set(&Eng_Global->instances[PLAYER1].entflags,ENTFLAG_GRAV_LIFT_STATE,true);
    if (o->velocity.y < e->offStrengthFactor) {
        float yForce = e->offStrengthFactor - o->velocity.y;
        if (initial || e->initialBurstFinished > Eng_Global->pauseRelativeTime) yForce *= 2.0f;
        AddForce(other,(Vector3){0.0f,yForce,0.0f},false);
    }
}

void GravityLiftOnTriggerEnter(uint16_t self, uint16_t other) {
    Eng_Global->instances[self].initialBurstFinished = Eng_Global->pauseRelativeTime + 1.0f;
    if (Eng_Global->instances[self].active) GravityLiftOnForce(self,other,true);
    else GravityLiftOffForce(self,other,true);
}

void GravityLiftOnTriggerStay(uint16_t self, uint16_t other) {
    if (Eng_Global->instances[self].active) GravityLiftOnForce(self,other,false);
    else GravityLiftOffForce(self,other,false);
}

void GravityLiftToggle(uint16_t self) { Eng_Global->instances[self].active = !Eng_Global->instances[self].active; }
